quickinput-sample/src/promptCommandWithHistory.ts

```diff
@@ -0,0 +1,104 @@
+import { window, Disposable, Memento } from 'vscode';
+import { QuickPickItem } from 'vscode';
+
+export const historyPath = `${process.env.HOME}/.vscode-cmd-history`;
+
+const HISTORY_KEY = 'HISTORY_KEY';
+
+/**
+ * A command prompt with history
+ * 
+ */
+export const getPromptCommand = (memo: Memento) => {
+	const pickCommand = getPickCommand(memo);
+	// §TODO: cache promptCommand!
+	return async function promptCommand() {
+		const command = await pickCommand();
+		if (command) {
+			window.showInformationMessage(`You picked the following command: '${command}'`)
+		}
+	}
+}
+
+abstract class CommandItem implements QuickPickItem {
+	public label: string;
+	public description?: string;
+	public abstract type: string;
+	constructor(label: string, description?: string) {
+		this.label = label;
+		this.description = description;
+	}
+}
+class HistoryItem extends CommandItem {
+	public type: string;
+	constructor(label: string, description?: string) {
+		super(label, description);
+		this.type = 'history'
+	}
+}
+class InputItem extends CommandItem {
+	public type: string;
+	constructor(public label: string) {
+		super(label, '(current input)');
+		this.type = 'input';
+	};
+}
+
+const getPickCommand = (memo: Memento) => async function pickCommand() {
+	const disposables: Disposable[] = [];
+	try {
+		return await new Promise<string | undefined>((resolve, reject) => {
+			const input = window.createQuickPick<CommandItem>();
+			input.placeholder = 'Type a command';
+			input.items = [];
+
+			const updateQuickPick = (value?: string): void => {
+				if (value === input.value) return;
+				if (!value) {
+					if (input.items[0] && input.items[0].type === 'input')
+					input.items = input.items.slice(1);
+					return;
+				}
+				if (input.items[0] && input.items[0].type === 'input') {
```

Comment:
Instead of having to distinguish 'input' from 'history' here, you could move the `historyItems` array to the beginning of this function and always use that to set the `input.items`. Would that simplify the code in this function?